experiment_prototype/scan_classes/sequences.py

```diff
@@ -104,203 +82,143 @@ def __init__(self, seqn_keys, sequence_slice_dict, sequence_interface, transmit_
+                phased_samps_for_beams = np.outer(phase_shift.flatten(), basic_samples)
+
+                # beams by antenna by samples
+                phased_samps_for_beams = phased_samps_for_beams.reshape(phase_shift.shape + basic_samples.shape)
+
+                # zero out the antennas not being used.
+                temp = np.zeros_like(phased_samps_for_beams, dtype=phased_samps_for_beams.dtype)
+                temp[:,exp_slice['tx_antennas'],:] = phased_samps_for_beams[:,exp_slice['tx_antennas'],:]
+                phased_samps_for_beams = temp[:,exp_slice['tx_antennas'],:]
+
+                self.basic_slice_pulses[slice_id] = phased_samps_for_beams
+            else:
+                self.basic_slice_pulses[slice_id] = []
+
+            for pulse_time in exp_slice['pulse_sequence']:
+                pulse_timing_us = pulse_time * exp_slice['tau_spacing'] + exp_slice['seqoffset']
+
+                single_pulse_timing.append({'start_time_us' : pulse_timing_us,
+                                            'pulse_len_us' : exp_slice['pulse_len'],
+                                            'slice_id' : slice_id})
+
+
+        single_pulse_timing = sorted(single_pulse_timing, key=lambda d: d['start_time_us'])
+
+        def make_pulse_dict(pulse_timing_info):
+            return {'start_time_us' : pulse_timing_info['start_time_us'],
+                      'total_pulse_len' : pulse_timing_info['pulse_len_us'],
+                      'component_info' : [pulse_timing_info]
+                    }
+
+        pulse_data = make_pulse_dict(single_pulse_timing[0])
+        combined_pulses_metadata = []
+
+        # determine where pulses occur in the sequence. This will be important if there are overlaps
+        for pulse_time in single_pulse_timing[1:]:
+            pulse_timing_us = pulse_time['start_time_us']
+            pulse_len_us = pulse_time['pulse_len_us']
 
-        # TODO: use number of frequencies to determine power output of each frequency (1/2, 1/3)
+            cp_timing_us = pulse_data['start_time_us']
+            cp_pulse_len_us = pulse_data['total_pulse_len']
 
+            min_sep = self.transmit_metadata['minimum_pulse_separation']
+            if pulse_timing_us < cp_timing_us + cp_pulse_len_us + min_sep:
+                new_pulse_len = pulse_timing_us - cp_timing_us + pulse_len_us
 
-        # TODO add in seqoffset value from the slice to pulse timing!!
-        pulses = []
-        # Getting a list of pulses, where pulse is a dictionary
+                pulse_data['total_pulse_len'] = new_pulse_len
+                pulse_data['component_info'].append(pulse_time)
+            else:
+                combined_pulses_metadata.append(pulse_data)
+                pulse_data = make_pulse_dict(pulse_time)
+
+        combined_pulses_metadata.append(pulse_data)
+
+        # normalize all pulses to max usrp dac amplitude.
+        power_divider = max([len(p['component_info']) for p in combined_pulses_metadata])
```

Comment:
I agree, although i don't think this should affect cutlass. I think I asked, and the ramps should not overlap, so these pulses won't be combined. 

I think this will be a problem if you transmit different signals on different antennas though for example